servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/JavaNetSoTimeoutHttpConnectionFilter.java

```diff
@@ -167,24 +173,101 @@ public Single<StreamingHttpResponse> request(final StreamingHttpRequest request)
+                                        Duration timeout, Executor timeoutExecutor) {
+            this.delegate = delegate;
+            this.requestComplete = requestComplete;
+            this.timeout = timeout;
+            this.timeoutExecutor = timeoutExecutor;
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            delegate.onSubscribe(() -> {
+                // We don't need to condition cancellation here on the `once()` call because the `DelayedCancellable`
+                // will already enforce idempotency of the cancel call, and it's fine if we're racing cancels with the
+                // results since cleanup is gated on the `once()` call.
+                once();
+                timeoutCancellable.cancel();
+                requestCancellable.cancel();
+            });
+            requestCancellable.delayedCancellable(cancellable);
+            timeoutCancellable.delayedCancellable(requestComplete.concat(Completable.never()
+                    .timeout(timeout, timeoutExecutor)).beforeOnError(this::handleInterruptions).subscribe());
+        }
+
+        private void handleInterruptions(Throwable t) {
+            if (once()) {
+                requestCancellable.cancel();
+                Throwable result = t;
+                // We can get a SocketTimeoutException waiting for a 100 Continue response.
+                if (t instanceof TimeoutException) {
+                result = newStacklessSocketTimeoutException(
+                        "Read timed out after " + timeout.toMillis() +
+                                "ms waiting for response meta-data")
+                        .initCause(t);
+                }
+                delegate.onError(result);
+            }
+        }
+
+        @Override
+        public void onSuccess(@Nullable StreamingHttpResponse result) {
+            if (once()) {
+                timeoutCancellable.cancel();
+                if (result != null) {
+                    result = result.transformMessageBody(p -> p.timeout(timeout, timeoutExecutor)
+                            .onErrorMap(TimeoutException.class, t -> newStacklessSocketTimeoutException(
+                                    "Read timed out after " + timeout.toMillis() +
+                                            "ms waiting for the next response payload body chunk")
+                                    .initCause(t)));
+                }
+                delegate.onSuccess(result);
+            } else {
+                if (result != null) {
+                    toSource(result.messageBody()).subscribe(CancelImmediatelySubscriber.INSTANCE);
+                }
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            if (once()) {
+                timeoutCancellable.cancel();
+                delegate.onError(t);
+            }
+        }
+
+        private boolean once() {
+            return 0 == onceUpdater.getAndSet(this, 1);
```

Comment:
this can be a simple cas, no need to loop inside `getAndSet` because the only possible transition is from 0 to 1

Reply:
The impl of this method:
```java
    public int getAndSet(T obj, int newValue) {
        int prev;
        do {
            prev = get(obj);
        } while (!compareAndSet(obj, prev, newValue));
        return prev;
    }
```
What are the additional guarantees that are important in this use-case besides CAS?

Reply:
Note that the impl is overridden in the only implementation that I'm aware of, `AtomicIntegerFieldUpdaterImpl`:
```
        public final int getAndSet(T obj, int newValue) {
            accessCheck(obj);
            return U.getAndSetInt(obj, offset, newValue);
        }
```
In this case it doesn't matter but `getAndSet` guarantees the write portion of the volatile happens-before relationship whereas CAS only guarantees the write if it succeeds.

Reply:
Thanks for sharing. I didn't know that there could be a difference between x86 and ARM, but in my head it always worked as it's interpreted by ARM. I was not surprised :)

In this case, we don't have any other state to worry about, and for consistency with other places we always use CAS in the same situations. It's easier to read and understand operators code if all places look identical, without the need to think why in this particular case it's implemented differently.

re: `AtomicIntegerFieldUpdaterImpl`
If you look inside `U`, it's implemented as:
```java
    @HotSpotIntrinsicCandidate
    public final int getAndSetInt(Object o, long offset, int newValue) {
        int v;
        do {
            v = getIntVolatile(o, offset);
        } while (!weakCompareAndSetInt(o, offset, v, newValue));
        return v;
    }
```
So, the optimization works only if intrinsics worked.

Reply:
While I still think `getAndSet` is better I changed it to a CAS to make progress. We can talk more about getAndSet in the future.